Firestore/core/src/firebase/firestore/core/sync_engine.h

```diff
@@ -0,0 +1,293 @@
+   * provided callback will be called once the write has been acked or
+   * rejected by the backend (or failed locally for any other reason).
+   */
+  void WriteMutations(std::vector<model::Mutation>&& mutations,
+                      util::StatusCallback callback);
+
+  /**
+   * Registers a user callback that is called when all pending mutations at the
+   * moment of calling are acknowledged .
+   */
+  void RegisterPendingWritesCallback(util::StatusCallback callback);
+
+  /**
+   * Runs the given transaction block up to retries times and then calls
+   * completion.
+   *
+   * @param retries The number of times to try before giving up.
+   * @param worker_queue The queue to dispatch sync engine calls to.
+   * @param update_callback The callback to call to execute the user's
+   * transaction.
+   * @param result_callback The callback to call when the transaction is
+   * finished or failed.
+   */
+  void Transaction(int retries,
+                   const std::shared_ptr<util::AsyncQueue>& worker_queue,
+                   core::TransactionUpdateCallback update_callback,
+                   core::TransactionResultCallback result_callback);
+
+  void HandleCredentialChange(const auth::User& user);
+
+  // Implements `RemoteStoreCallback`
+  void ApplyRemoteEvent(const remote::RemoteEvent& remote_event);
+  void HandleRejectedListen(model::TargetId target_id, util::Status error);
+  void HandleSuccessfulWrite(const model::MutationBatchResult& batch_result);
+  void HandleRejectedWrite(firebase::firestore::model::BatchId batchID,
+                           util::Status error);
+  void HandleOnlineStateChange(model::OnlineState online_state);
+  model::DocumentKeySet GetRemoteKeys(model::TargetId targetId) const;
+
+  // For tests only
+  std::map<model::DocumentKey, model::TargetId> GetCurrentLimboDocuments()
+      const {
+    // Return defensive copy
+    return limbo_targets_by_key_;
+  }
+
+ private:
+  /**
+   * QueryView contains all of the info that SyncEngine needs to track for a
+   * particular query and view.
+   */
+  class QueryView {
+   public:
+    QueryView(Query query,
+              model::TargetId target_id,
+              nanopb::ByteString resume_token,
+              View view)
+        : query_(std::move(query)),
+          target_id_(target_id),
+          resume_token_(std::move(resume_token)),
+          view_(std::move(view)) {
+    }
+
+    const Query& query() const {
+      return query_;
+    }
+
+    /**
+     * The target ID created by the client that is used in the watch stream to
+     * identify this query.
+     */
+    model::TargetId target_id() const {
+      return target_id_;
+    }
+
+    /**
+     * An identifier from the datastore backend that indicates the last state of
+     * the results that was received. This can be used to indicate where to
+     * continue receiving new doc changes for the query.
+     */
+    const nanopb::ByteString resume_token() const {
```

Comment:
This should return either a non-`const` value or a const reference (probably the latter).